### DISCUSSION TITLE
Is dagger suitable for me?

### DISCUSSION THREAD

**Discussion by cathaysia** [dagger/dagger#6363] [Help] [answered]
hi. Hello everyone. I'm currently looking for a cicd tool that will suit my needs. I hope someone can give me some opinions.

My team currently uses Jenkins as the cicd tool, but you know what. Jenkins is not very convenient to use. In the past week, I spent a lot of effort getting `jekins in docker` and `docker in jenkins` running. Soon, I will have to write many, many scripts (or Makefiles) on it. It makes me miss the fun days of using Github Action.

I have been trying to replace jenkins with a better CICD tool, but I still haven't found a more suitable one. GOCD looks great, but I haven't looked into it in depth yet. buildbot looks great and is used by many large projects. But I don't like python that much (although it helped me a lot). Dagger looks similar to buildbot, and it also needs to use code to control the build process.

My team's current environment and work is as follows:

- The code warehouse is located on the public network, while cicd is located on the intranet. So webhooks cannot be used (but the code repository itself also has a cicd system and can execute scripts to call my own hosted cicd.).
- Need to use cicd to execute tests
- Need to use cicd to build code and generate sdk.

Overall, I would like cicd to:

- Ability to execute tests using Docker.
- Ability to inject ssh keys into Docker to clone repositories.
- Ability to manage caching (even failed tests are cached since they are downloaded from the network).
- Able to manage artifacts (or upload to other artifact managers)
- Can be triggered via API.
- Ability to return test results.
- A GUI interface (as long as it can display the running status).

It would be better if it has the following functions:
- Able to use matrix to perform tests in different environments
- Able to use matrix to generate sdk in different environments.
- Skip the test if the file has not changed.
- Ability to expand horizontally.
- Convenient for debugging.
- Ability to display my own generated HTML in the browser.
- Easy to deploy and use. (I hope a week will be enough time for me to switch over my existing tests)

**Comment by samalba** [comment, marked as answer]
Hello @cathaysia,

Before digging into the details, yes, you can easily execute your tests, build and run docker containers, manage artifacts, call APIs, etc... from a Dagger pipeline. You can then run this pipeline either locally from your machine or from Jenkins or from any infrastructure. [Dagger Cloud](https://dagger.cloud/) can then allow you to visualize the running status (logs and errors) from a Web UI.

I would go over the following resources, in that order (relevant to the points you listed):

1. Video: [general overview Dagger](https://www.youtube.com/watch?v=HHhttaiHtm4) - also the rest of that channel has many short videos focusing on a particular use cases
2. Tutorial (and explanation of key concepts): [Dagger Quickstart](https://docs.dagger.io/quickstart)
3. Dagger Cookbook (https://docs.dagger.io/cookbook)
  a. In particular, [the section about Build](https://docs.dagger.io/cookbook#builds), it has lots of relevant sub-sections around building docker images, matrix of multi-arch builds, etc...

This should cover most (if not all) points you listed. If you need help while experimenting, I invite you to join our [discord server](https://discord.gg/ufnyBtc8uY), the whole Dagger Team and Community is connected and ready to help, we can easily organize a quick video call where we show you live some relevant examples.

Curious to know what you build next!

**Reply by cathaysia** [reply]
Thank you very much, I will learn more about dagger from the material you provided。 :)
